Better error message when UNAUTHORIZED

UNAUTHORIZED is a common failure, and the current error message is very general to set a credential helper or credentials in a build config file.

`[ERROR] Failed to execute goal com.google.cloud.tools:jib-maven-plugin:1.0.2:build (default-cli) on project hello-spring-boot: Build image failed, perhaps you should set a credential helper name with the configuration '<to><credHelper>' or set credentials for 'gcr.io' in your Maven settings: Unauthorized for ...: 401 Unauthorized
[ERROR] {"errors":[{"code":"UNAUTHORIZED","message":"You don't have the needed permissions to perform this operation, and you may have invalid credentials. To authenticate your request, follow the steps in: https://cloud.google.com/container-registry/docs/advanced-authentication"}]}`

We can have a detailed FAQ entry and put the link in the error message.

--- a/docs/faq.md
+++ b/docs/faq.md
@@ -20,6 +20,7 @@ If a question you have is not answered below, please [submit an issue](/../../is
 [How can I inspect the image Jib built?](#how-can-i-inspect-the-image-jib-built)\
 [I am seeing `ImagePullBackoff` on my pods.](#i-am-seeing-imagepullbackoff-on-my-pods-in-minikube)\
 [How do I configure a proxy?](#how-do-i-configure-a-proxy)\
+[What should I do when the registry responds with Forbidden or DENIED?](#what-should-i-do-when-the-registry-responds-with-forbidden-or-denied)\
 [What should I do when the registry responds with UNAUTHORIZED?](#what-should-i-do-when-the-registry-responds-with-unauthorized)\
 [How can I diagnose problems pulling or pushing from remote registries?](#how-can-i-diagnose-problems-pulling-or-pushing-from-remote-registries)\
 [How can I examine network traffic?](#how-can-i-examine-network-traffic)\
@@ -344,6 +345,12 @@ See more at [Using Google Container Registry (GCR) with Minikube](https://ryanes
 
 Jib currently requires configuring your build tool to use the appropriate [Java networking properties](https://docs.oracle.com/javase/8/docs/api/java/net/doc-files/net-properties.html) (`https.proxyHost`, `https.proxyPort`, `https.proxyUser`, `https.proxyPassword`).
 
+### What should I do when the registry responds with Forbidden or DENIED?
+
+If the registry returns "403 Forbidden" or `"code":"DENIED"`, it often means Jib successfully authenticated using credentials you provided but the credentials do not have permissions to pull or push images. Make sure your account/role has the permissions to do the operation.
+
+However, depending on registry implementations, it is also possible that the registry actually meant you are not authenticated. Sometimes, it may be due to Jib picking up different credentials, credentials incorrectly configured, or typos in auth configuration. Or, it may be that your credentials work for pulling the base image but not for pushing. See [What should I do when the registry responds with UNAUTHORIZED?](#what-should-i-do-when-the-registry-responds-with-unauthorized) to ensure you have set up credentials correctly.
+
 ### What should I do when the registry responds with UNAUTHORIZED?
 
 If the registry returns "401 Unauthorized" or `"code":"UNAUTHORIZED"`, it is often due to credential misconfiguration. Examples:
@@ -355,8 +362,10 @@ If the registry returns "401 Unauthorized" or `"code":"UNAUTHORIZED"`, it is oft
       - Configuring credential helpers: [`<from/to><credHelper>`](https://github.com/GoogleContainerTools/jib/tree/master/jib-maven-plugin#using-docker-credential-helpers) for Maven / [`from/to.credHelper`](https://github.com/GoogleContainerTools/jib/tree/master/jib-gradle-plugin#using-docker-credential-helpers) for Gradle
       - Specific credentials (not recommend): [`<from/to><auth><username>/<password>`](https://github.com/GoogleContainerTools/jib/tree/master/jib-maven-plugin#using-specific-credentials) or in [`settings.xml`](https://github.com/GoogleContainerTools/jib/tree/master/jib-maven-plugin#using-maven-settings) for Maven / [`from/to.auth.username/password`](https://github.com/GoogleContainerTools/jib/tree/master/jib-gradle-plugin#using-specific-credentials) for Gradle
       - These parameters can also be set through properties: [Maven](https://github.com/GoogleContainerTools/jib/tree/master/jib-maven-plugin#system-properties) / [Gradle](https://github.com/GoogleContainerTools/jib/tree/master/jib-gradle-plugin#system-properties)
+* There exist different auth configurations in multiple places, and Jib is not picking up the auth information you are working on.
 * You configured a credential helper, but the helper is not on `$PATH`. This is especially common when running Jib inside IDE where the IDE binary is launched directly from an OS menu and does not have access to your shell's environment.
 * Typos in username, password, image names, or registry names.
+* You are using a private registry without HTTPS. See [How can I diagnose problems pulling or pushing from remote registries?](#how-can-i-diagnose-problems-pulling-or-pushing-from-remote-registries) if this is the case.
 
 If you encounter issues interacting with a registry other than `UNAUTHORIZED`, check ["How can I diagnose problems pulling or pushing from remote registries?"](#how-can-i-diagnose-problems-pulling-or-pushing-from-remote-registries).
 

--- a/docs/faq.md
+++ b/docs/faq.md
@@ -347,13 +347,13 @@ Jib currently requires configuring your build tool to use the appropriate [Java 
 
 ### What should I do when the registry responds with Forbidden or DENIED?
 
-If the registry returns "403 Forbidden" or `"code":"DENIED"`, it often means Jib successfully authenticated using credentials you provided but the credentials do not have permissions to pull or push images. Make sure your account/role has the permissions to do the operation.
+If the registry returns `403 Forbidden` or `"code":"DENIED"`, it often means Jib successfully authenticated using your credentials but the credentials do not have permissions to pull or push images. Make sure your account/role has the permissions to do the operation.
 
-However, depending on registry implementations, it is also possible that the registry actually meant you are not authenticated. Sometimes, it may be due to Jib picking up different credentials, credentials incorrectly configured, or typos in auth configuration. Or, it may be that your credentials work for pulling the base image but not for pushing. See [What should I do when the registry responds with UNAUTHORIZED?](#what-should-i-do-when-the-registry-responds-with-unauthorized) to ensure you have set up credentials correctly.
+Depending on registry implementations, it is also possible that the registry actually meant you are not authenticated. See [What should I do when the registry responds with UNAUTHORIZED?](#what-should-i-do-when-the-registry-responds-with-unauthorized) to ensure you have set up credentials correctly.
 
 ### What should I do when the registry responds with UNAUTHORIZED?
 
-If the registry returns "401 Unauthorized" or `"code":"UNAUTHORIZED"`, it is often due to credential misconfiguration. Examples:
+If the registry returns `401 Unauthorized` or `"code":"UNAUTHORIZED"`, it is often due to credential misconfiguration. Examples:
 
 * You did not configure auth information in the default places where Jib searches.
    - `$HOME/.docker/config.json`, [one of the configuration files](https://docs.docker.com/engine/reference/commandline/cli/#configuration-files) for the `docker` command line tool. See [configuration files document](https://docs.docker.com/engine/reference/commandline/cli/#configuration-files), [credential store](https://docs.docker.com/engine/reference/commandline/login/#credentials-store) and [credential helper](https://docs.docker.com/engine/reference/commandline/login/#credential-helpers) sections, and [this](https://github.com/GoogleContainerTools/jib/issues/101) for how to configure auth. For example, you can do `docker login` to save auth in `config.json`, but it is often recommended to configure a credential helper (also configurable in `config.json`).
@@ -362,10 +362,10 @@ If the registry returns "401 Unauthorized" or `"code":"UNAUTHORIZED"`, it is oft
       - Configuring credential helpers: [`<from/to><credHelper>`](https://github.com/GoogleContainerTools/jib/tree/master/jib-maven-plugin#using-docker-credential-helpers) for Maven / [`from/to.credHelper`](https://github.com/GoogleContainerTools/jib/tree/master/jib-gradle-plugin#using-docker-credential-helpers) for Gradle
       - Specific credentials (not recommend): [`<from/to><auth><username>/<password>`](https://github.com/GoogleContainerTools/jib/tree/master/jib-maven-plugin#using-specific-credentials) or in [`settings.xml`](https://github.com/GoogleContainerTools/jib/tree/master/jib-maven-plugin#using-maven-settings) for Maven / [`from/to.auth.username/password`](https://github.com/GoogleContainerTools/jib/tree/master/jib-gradle-plugin#using-specific-credentials) for Gradle
       - These parameters can also be set through properties: [Maven](https://github.com/GoogleContainerTools/jib/tree/master/jib-maven-plugin#system-properties) / [Gradle](https://github.com/GoogleContainerTools/jib/tree/master/jib-gradle-plugin#system-properties)
-* There exist different auth configurations in multiple places, and Jib is not picking up the auth information you are working on.
+* Different auth configurations exist in multiple places, and Jib is not picking up the auth information you are working on.
 * You configured a credential helper, but the helper is not on `$PATH`. This is especially common when running Jib inside IDE where the IDE binary is launched directly from an OS menu and does not have access to your shell's environment.
 * Typos in username, password, image names, or registry names.
-* You are using a private registry without HTTPS. See [How can I diagnose problems pulling or pushing from remote registries?](#how-can-i-diagnose-problems-pulling-or-pushing-from-remote-registries) if this is the case.
+* You are using a private registry without HTTPS. See [How can I diagnose problems pulling or pushing from remote registries?](#how-can-i-diagnose-problems-pulling-or-pushing-from-remote-registries).
 
 If you encounter issues interacting with a registry other than `UNAUTHORIZED`, check ["How can I diagnose problems pulling or pushing from remote registries?"](#how-can-i-diagnose-problems-pulling-or-pushing-from-remote-registries).
 

--- a/jib-gradle-plugin/src/main/java/com/google/cloud/tools/jib/gradle/GradleHelpfulSuggestionsBuilder.java
+++ b/jib-gradle-plugin/src/main/java/com/google/cloud/tools/jib/gradle/GradleHelpfulSuggestionsBuilder.java
@@ -66,12 +66,8 @@ class GradleHelpfulSuggestionsBuilder {
         "gradle clean",
         baseImageReference,
         !isCredHelperDefinedForBaseImage && !baseImageHasConfiguredCredentials,
-        "from.credHelper",
-        ignored -> "from.auth",
         targetImageReference,
         !isCredHelperDefinedForTargetImage && !targetImageHasConfiguredCredentials,
-        "to.credHelper",
-        ignored -> "to.auth",
         "jib.to.image",
         "--image",
         "build.gradle");

--- a/jib-maven-plugin/src/main/java/com/google/cloud/tools/jib/maven/MavenHelpfulSuggestionsBuilder.java
+++ b/jib-maven-plugin/src/main/java/com/google/cloud/tools/jib/maven/MavenHelpfulSuggestionsBuilder.java
@@ -18,14 +18,10 @@ package com.google.cloud.tools.jib.maven;
 
 import com.google.cloud.tools.jib.image.ImageReference;
 import com.google.cloud.tools.jib.plugins.common.HelpfulSuggestions;
-import java.util.function.Function;
 import javax.annotation.Nullable;
 
 /** Builder for Maven-specific {@link HelpfulSuggestions}. */
 class MavenHelpfulSuggestionsBuilder {
-
-  private static final Function<String, String> AUTH_CONFIGURATION_SUGGESTION =
-      registry -> "set credentials for '" + registry + "' in your Maven settings";
 
   private final String messagePrefix;
   private final JibPluginConfiguration jibPluginConfiguration;
@@ -78,12 +74,8 @@ class MavenHelpfulSuggestionsBuilder {
         "mvn clean",
         baseImageReference,
         !isCredHelperDefinedForBaseImage && !baseImageHasConfiguredCredentials,
-        "<from><credHelper>",
-        AUTH_CONFIGURATION_SUGGESTION,
         targetImageReference,
         !isCredHelperDefinedForTargetImage && !targetImageHasConfiguredCredentials,
-        "<to><credHelper>",
-        AUTH_CONFIGURATION_SUGGESTION,
         "<to><image>",
         "-Dimage",
         "pom.xml");

--- a/jib-plugins-common/src/main/java/com/google/cloud/tools/jib/plugins/common/HelpfulSuggestions.java
+++ b/jib-plugins-common/src/main/java/com/google/cloud/tools/jib/plugins/common/HelpfulSuggestions.java
@@ -19,7 +19,6 @@ package com.google.cloud.tools.jib.plugins.common;
 import com.google.cloud.tools.jib.image.ImageReference;
 import com.google.common.base.Preconditions;
 import java.nio.file.Path;
-import java.util.function.Function;
 import javax.annotation.Nullable;
 
 /** Builds messages that provides suggestions on how to fix the error. */
@@ -110,12 +109,8 @@ public class HelpfulSuggestions {
   private final String clearCacheCommand;
   @Nullable private final ImageReference baseImageReference;
   private final boolean noCredentialsDefinedForBaseImage;
-  private final String baseImageCredHelperConfiguration;
-  private final Function<String, String> baseImageAuthConfiguration;
   @Nullable private final ImageReference targetImageReference;
   private final boolean noCredentialsDefinedForTargetImage;
-  private final String targetImageCredHelperConfiguration;
-  private final Function<String, String> targetImageAuthConfiguration;
   private final String toImageConfiguration;
   private final String buildConfigurationFilename;
   private final String toImageFlag;
@@ -128,17 +123,9 @@ public class HelpfulSuggestions {
    * @param baseImageReference the base image reference
    * @param noCredentialsDefinedForBaseImage {@code true} if no credentials were defined for the
    *     base image; {@code false} otherwise
-   * @param baseImageCredHelperConfiguration the configuration defining the credential helper name
-   *     for the base image
-   * @param baseImageAuthConfiguration the way to define raw credentials for the base image - takes
-   *     the base image registry as an argument
    * @param targetImageReference the target image reference
    * @param noCredentialsDefinedForTargetImage {@code true} if no credentials were defined for the
-   *     base image; {@code false} otherwise
-   * @param targetImageCredHelperConfiguration the configuration defining the credential helper name
-   *     for the target image
-   * @param targetImageAuthConfiguration the way to define raw credentials for the target image -
-   *     takes the target image registry as an argument
+   *     target image; {@code false} otherwise
    * @param toImageConfiguration the configuration defining the target image
    * @param toImageFlag the commandline flag used to set the target image
    * @param buildConfigurationFilename the filename of the build configuration
@@ -148,12 +135,8 @@ public class HelpfulSuggestions {
       String clearCacheCommand,
       @Nullable ImageReference baseImageReference,
       boolean noCredentialsDefinedForBaseImage,
-      String baseImageCredHelperConfiguration,
-      Function<String, String> baseImageAuthConfiguration,
       @Nullable ImageReference targetImageReference,
       boolean noCredentialsDefinedForTargetImage,
-      String targetImageCredHelperConfiguration,
-      Function<String, String> targetImageAuthConfiguration,
       String toImageConfiguration,
       String toImageFlag,
       String buildConfigurationFilename) {
@@ -161,12 +144,8 @@ public class HelpfulSuggestions {
     this.clearCacheCommand = clearCacheCommand;
     this.baseImageReference = baseImageReference;
     this.noCredentialsDefinedForBaseImage = noCredentialsDefinedForBaseImage;
-    this.baseImageCredHelperConfiguration = baseImageCredHelperConfiguration;
-    this.baseImageAuthConfiguration = baseImageAuthConfiguration;
     this.targetImageReference = targetImageReference;
     this.noCredentialsDefinedForTargetImage = noCredentialsDefinedForTargetImage;
-    this.targetImageCredHelperConfiguration = targetImageCredHelperConfiguration;
-    this.targetImageAuthConfiguration = targetImageAuthConfiguration;
     this.toImageConfiguration = toImageConfiguration;
     this.buildConfigurationFilename = buildConfigurationFilename;
     this.toImageFlag = toImageFlag;
@@ -193,27 +172,37 @@ public class HelpfulSuggestions {
   }
 
   public String forHttpStatusCodeForbidden(String imageReference) {
-    return suggest("make sure you have permissions for " + imageReference);
+    return suggest(
+        "make sure you have permissions for "
+            + imageReference
+            + " and set correct credentials for the registry. See "
+            + "https://github.com/GoogleContainerTools/jib/blob/master/docs/faq.md#what-should-i-do-when-the-registry-responds-with-forbidden-or-denied for help");
   }
 
   public String forNoCredentialsDefined(String registry, String repository) {
     Preconditions.checkNotNull(baseImageReference);
     Preconditions.checkNotNull(targetImageReference);
-    if (noCredentialsDefinedForBaseImage
-        && registry.equals(baseImageReference.getRegistry())
-        && repository.equals(baseImageReference.getRepository())) {
-      return forNoCredentialHelpersDefined(
-          baseImageCredHelperConfiguration, baseImageAuthConfiguration.apply(registry));
+
+    final String unauthorizedRegistry;
+    if (registry.equals(baseImageReference.getRegistry())
+        && repository.equals(baseImageReference.getRepository())
+        && noCredentialsDefinedForBaseImage) {
+      unauthorizedRegistry = baseImageReference.toString();
+
+    } else if (registry.equals(targetImageReference.getRegistry())
+        && repository.equals(targetImageReference.getRepository())
+        && noCredentialsDefinedForTargetImage) {
+      unauthorizedRegistry = targetImageReference.toString();
+
+    } else {
+      unauthorizedRegistry = registry;
     }
-    if (noCredentialsDefinedForTargetImage
-        && registry.equals(targetImageReference.getRegistry())
-        && repository.equals(targetImageReference.getRepository())) {
-      return forNoCredentialHelpersDefined(
-          targetImageCredHelperConfiguration, targetImageAuthConfiguration.apply(registry));
-    }
-    // Credential helper probably was not configured correctly or did not have the necessary
-    // credentials.
-    return forCredentialsNotCorrect(registry);
+
+    return suggest(
+        "make sure your credentials for '"
+            + unauthorizedRegistry
+            + "' are set up correctly. See "
+            + "https://github.com/GoogleContainerTools/jib/blob/master/docs/faq.md#what-should-i-do-when-the-registry-responds-with-unauthorized for help");
   }
 
   public String forCredentialsNotSent() {
@@ -250,18 +239,5 @@ public class HelpfulSuggestions {
    */
   public String suggest(String suggestion) {
     return suggest(messagePrefix, suggestion);
-  }
-
-  private String forNoCredentialHelpersDefined(
-      String credHelperConfiguration, String authConfiguration) {
-    return suggest(
-        "set a credential helper name with the configuration '"
-            + credHelperConfiguration
-            + "' or "
-            + authConfiguration);
-  }
-
-  private String forCredentialsNotCorrect(String registry) {
-    return suggest("make sure your credentials for '" + registry + "' are set up correctly");
   }
 }

--- a/jib-plugins-common/src/main/java/com/google/cloud/tools/jib/plugins/common/HelpfulSuggestions.java
+++ b/jib-plugins-common/src/main/java/com/google/cloud/tools/jib/plugins/common/HelpfulSuggestions.java
@@ -175,7 +175,7 @@ public class HelpfulSuggestions {
     return suggest(
         "make sure you have permissions for "
             + imageReference
-            + " and set correct credentials for the registry. See "
+            + " and set correct credentials. See "
             + "https://github.com/GoogleContainerTools/jib/blob/master/docs/faq.md#what-should-i-do-when-the-registry-responds-with-forbidden-or-denied for help");
   }
 
@@ -183,24 +183,24 @@ public class HelpfulSuggestions {
     Preconditions.checkNotNull(baseImageReference);
     Preconditions.checkNotNull(targetImageReference);
 
-    final String unauthorizedRegistry;
+    final String unauthorizedEntity;
     if (registry.equals(baseImageReference.getRegistry())
         && repository.equals(baseImageReference.getRepository())
         && noCredentialsDefinedForBaseImage) {
-      unauthorizedRegistry = baseImageReference.toString();
+      unauthorizedEntity = baseImageReference.toString();
 
     } else if (registry.equals(targetImageReference.getRegistry())
         && repository.equals(targetImageReference.getRepository())
         && noCredentialsDefinedForTargetImage) {
-      unauthorizedRegistry = targetImageReference.toString();
+      unauthorizedEntity = targetImageReference.toString();
 
     } else {
-      unauthorizedRegistry = registry;
+      unauthorizedEntity = registry;
     }
 
     return suggest(
         "make sure your credentials for '"
-            + unauthorizedRegistry
+            + unauthorizedEntity
             + "' are set up correctly. See "
             + "https://github.com/GoogleContainerTools/jib/blob/master/docs/faq.md#what-should-i-do-when-the-registry-responds-with-unauthorized for help");
   }

--- a/jib-plugins-common/src/main/java/com/google/cloud/tools/jib/plugins/common/HelpfulSuggestions.java
+++ b/jib-plugins-common/src/main/java/com/google/cloud/tools/jib/plugins/common/HelpfulSuggestions.java
@@ -183,19 +183,15 @@ public class HelpfulSuggestions {
     Preconditions.checkNotNull(baseImageReference);
     Preconditions.checkNotNull(targetImageReference);
 
-    final String unauthorizedEntity;
+    String unauthorizedEntity = registry;
     if (registry.equals(baseImageReference.getRegistry())
         && repository.equals(baseImageReference.getRepository())
         && noCredentialsDefinedForBaseImage) {
       unauthorizedEntity = baseImageReference.toString();
-
     } else if (registry.equals(targetImageReference.getRegistry())
         && repository.equals(targetImageReference.getRepository())
         && noCredentialsDefinedForTargetImage) {
       unauthorizedEntity = targetImageReference.toString();
-
-    } else {
-      unauthorizedEntity = registry;
     }
 
     return suggest(

--- a/jib-plugins-common/src/test/java/com/google/cloud/tools/jib/plugins/common/ConfigurationPropertyValidatorTest.java
+++ b/jib-plugins-common/src/test/java/com/google/cloud/tools/jib/plugins/common/ConfigurationPropertyValidatorTest.java
@@ -129,12 +129,8 @@ public class ConfigurationPropertyValidatorTest {
             "",
             mockImageReference,
             false,
-            "",
-            unused -> "",
             mockImageReference,
             false,
-            "",
-            unused -> "",
             "to",
             "--to",
             "build.txt");

--- a/jib-plugins-common/src/test/java/com/google/cloud/tools/jib/plugins/common/HelpfulSuggestionsTest.java
+++ b/jib-plugins-common/src/test/java/com/google/cloud/tools/jib/plugins/common/HelpfulSuggestionsTest.java
@@ -51,7 +51,7 @@ public class HelpfulSuggestionsTest {
         "messagePrefix, perhaps you should check that 'cacheDirectory' is not used by another application or set the `jib.useOnlyProjectCache` system property",
         TEST_HELPFUL_SUGGESTIONS.forCacheDirectoryNotOwned(Paths.get("cacheDirectory")));
     Assert.assertEquals(
-        "messagePrefix, perhaps you should make sure you have permissions for imageReference and set correct credentials for the registry. See https://github.com/GoogleContainerTools/jib/blob/master/docs/faq.md#what-should-i-do-when-the-registry-responds-with-forbidden-or-denied for help",
+        "messagePrefix, perhaps you should make sure you have permissions for imageReference and set correct credentials. See https://github.com/GoogleContainerTools/jib/blob/master/docs/faq.md#what-should-i-do-when-the-registry-responds-with-forbidden-or-denied for help",
         TEST_HELPFUL_SUGGESTIONS.forHttpStatusCodeForbidden("imageReference"));
     Assert.assertEquals(
         "messagePrefix, perhaps you should make sure your credentials for 'baseregistry/baserepository' are set up correctly. See https://github.com/GoogleContainerTools/jib/blob/master/docs/faq.md#what-should-i-do-when-the-registry-responds-with-unauthorized for help",

--- a/jib-plugins-common/src/test/java/com/google/cloud/tools/jib/plugins/common/HelpfulSuggestionsTest.java
+++ b/jib-plugins-common/src/test/java/com/google/cloud/tools/jib/plugins/common/HelpfulSuggestionsTest.java
@@ -30,12 +30,8 @@ public class HelpfulSuggestionsTest {
           "clearCacheCommand",
           ImageReference.of("baseregistry", "baserepository", null),
           true,
-          "baseImageCredHelperConfiguration",
-          registry -> "baseImageAuthConfiguration " + registry,
           ImageReference.of("targetregistry", "targetrepository", null),
           false,
-          "targetImageCredHelperConfiguration",
-          registry -> "targetImageAuthConfiguration " + registry,
           "toProperty",
           "toFlag",
           "buildFile");
@@ -55,13 +51,13 @@ public class HelpfulSuggestionsTest {
         "messagePrefix, perhaps you should check that 'cacheDirectory' is not used by another application or set the `jib.useOnlyProjectCache` system property",
         TEST_HELPFUL_SUGGESTIONS.forCacheDirectoryNotOwned(Paths.get("cacheDirectory")));
     Assert.assertEquals(
-        "messagePrefix, perhaps you should make sure you have permissions for imageReference",
+        "messagePrefix, perhaps you should make sure you have permissions for imageReference and set correct credentials for the registry. See https://github.com/GoogleContainerTools/jib/blob/master/docs/faq.md#what-should-i-do-when-the-registry-responds-with-forbidden-or-denied for help",
         TEST_HELPFUL_SUGGESTIONS.forHttpStatusCodeForbidden("imageReference"));
     Assert.assertEquals(
-        "messagePrefix, perhaps you should set a credential helper name with the configuration 'baseImageCredHelperConfiguration' or baseImageAuthConfiguration baseregistry",
+        "messagePrefix, perhaps you should make sure your credentials for 'baseregistry/baserepository' are set up correctly. See https://github.com/GoogleContainerTools/jib/blob/master/docs/faq.md#what-should-i-do-when-the-registry-responds-with-unauthorized for help",
         TEST_HELPFUL_SUGGESTIONS.forNoCredentialsDefined("baseregistry", "baserepository"));
     Assert.assertEquals(
-        "messagePrefix, perhaps you should make sure your credentials for 'targetregistry' are set up correctly",
+        "messagePrefix, perhaps you should make sure your credentials for 'targetregistry' are set up correctly. See https://github.com/GoogleContainerTools/jib/blob/master/docs/faq.md#what-should-i-do-when-the-registry-responds-with-unauthorized for help",
         TEST_HELPFUL_SUGGESTIONS.forNoCredentialsDefined("targetregistry", "targetrepository"));
     Assert.assertEquals(
         "messagePrefix, perhaps you should add a `mainClass` configuration to plugin",

--- a/jib-plugins-common/src/test/java/com/google/cloud/tools/jib/plugins/common/JibBuildRunnerTest.java
+++ b/jib-plugins-common/src/test/java/com/google/cloud/tools/jib/plugins/common/JibBuildRunnerTest.java
@@ -51,12 +51,8 @@ public class JibBuildRunnerTest {
           "clearCacheCommand",
           ImageReference.of("someregistry", "somerepository", null),
           false,
-          "baseImageCredHelperConfiguration",
-          registry -> "baseImageAuthConfiguration " + registry,
           ImageReference.of("toRegistry", "torepository", null),
           false,
-          "targetImageCredHelperConfiguration",
-          registry -> "targetImageAuthConfiguration " + registry,
           "toConfig",
           "toFlag",
           "buildFile");

